functional_programming_tests/test/with_functional_programming.cpp

```diff
@@ -0,0 +1,94 @@
+#include <gtest/gtest.h>
+
+#include <iostream>
+#include <vector>
+#include <cmath>
+#include <optional>
+
+// production_code.h/cc
+template<typename T>
+using Map = std::vector<std::vector<T>>;
```

Comment:
For this example I think this is ok, but I don't think we would normally store a "map" like this.

Reply:
I see. How is a "map" normally stored?

Reply:
Usually it's a 1-d array that has accessor functions that jump to the correct index. Separating the representation and the data storage. This may be a legacy practice but it may have practical value. Not sure.

Reply:
gpt says:

Yes, the assertion that maps or images are commonly stored as 1D arrays instead of `std::vector<std::vector<>>` in a C++ context is correct. There are a few reasons for this:

1. Performance: Using a 1D array to store a 2D map or image provides better cache locality, which helps improve the performance of the code. Accessing elements that are close to each other in memory (as they would be in a 1D array) is generally faster than accessing elements scattered throughout memory (as might be the case with a nested `std::vector`).

2. Memory overhead: A nested `std::vector` has additional memory overhead because each inner `std::vector` stores its size and capacity separately. In contrast, a 1D array has lower memory overhead, as it only needs to store the size and capacity once for the entire structure.

3. Simplicity: A 1D array is simpler to allocate and manage. For example, if you have a 2D array of dimensions `width` and `height`, you can easily calculate the index of an element `(x, y)` in a 1D array using the formula `index = y * width + x`. This makes it easier to pass the array to functions or libraries that expect a single block of memory.

4. Compatibility: Many libraries and APIs that work with images or 2D data expect a single block of memory (like a 1D array) rather than nested data structures. Using a 1D array allows for easier compatibility with these libraries and APIs.

While using a 1D array has these advantages, there are cases where a nested `std::vector` or another 2D container might be more appropriate, such as when the dimensions of the map or image are not fixed or when dealing with a sparse dataset. As with many programming decisions, the choice depends on the specific problem you're trying to solve and the trade-offs you're willing to make.

Reply:
Of course in a real application we would probably use an `Eigen::Matrix` of some sort